src/nvim/tag.c

```diff
@@ -1974,20 +1960,27 @@ find_tags (
     matches = NULL;
   match_count = 0;
   for (mtt = 0; mtt < MT_COUNT; ++mtt) {
-    for (int i = 0; i < ga_match[mtt].ga_len; ++i) {
-      mfp = ((struct match_found **)(ga_match[mtt].ga_data))[i];
-      if (matches == NULL)
-        xfree(mfp);
-      else {
-        /* To avoid allocating memory again we turn the struct
-         * match_found into a string.  For help the priority was not
-         * included in the length. */
-        memmove(mfp, mfp->match,
-            (size_t)(mfp->len + (help_only ? ML_HELP_LEN : 0)));
-        matches[match_count++] = (char_u *)mfp;
+    hashitem_T *hi;
+    long todo = (long)ht_match[mtt].ht_used;
+
+    for (hi = ht_match[mtt].ht_array; todo > 0; hi++) {
+      if (!HASHITEM_EMPTY(hi)) {
+        mfp = hi->hi_key;
+        if (matches == NULL) {
+          xfree(mfp);
+        } else {
+          // now change the TAG_SEP back to NUL
+          for (p = mfp; *p != NUL; p++) {
+            if (*p == TAG_SEP) {
+              *p = NUL;
+            }
+          }
```

Comment:
I'm not familiar with all the utility functions that we've added, and I wasn't looking for a function to handle this.